### DISCUSSION TITLE
Package a package binary with PyInstaller (awscli)

### DISCUSSION THREAD

**Discussion by RobinFrcd** [pyinstaller/pyinstaller#7153] [Help] [answered]
Hi,
I'd like to include the `aws` binary provided in the [`awscli` package](https://pypi.org/project/awscli/). I went through the [Understanding PyInstaller Hooks doc](https://pyinstaller.org/en/stable/hooks.html) and thought that `collect_dynamic_libs("awscli")` would help me here, but this function doesn't return anything ?

Is there an easy way to package the AWS CLI (for all systems) ? For example, un Unix systems it's located here: `sys.executable/../aws`

Thanks

**Comment by bwoodsend** [comment, marked as answer]
I'm guessing by _binary_ you actually mean the CLI entry point script `aws` in `sys.prefix/bin/aws`? That's not a binary file (unless you're on Windows) - it's just a Python file without a `.py` suffix. It's contents:
```python
#!/usr/local/bin/python
# Copyright 2012 Amazon.com, Inc. or its affiliates. All Rights Reserved.

# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at

#     http://aws.amazon.com/apache2.0/

# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.
import sys
import os

if os.environ.get('LC_CTYPE', '') == 'UTF-8':
    os.environ['LC_CTYPE'] = 'en_US.UTF-8'
import awscli.clidriver


def main():
    return awscli.clidriver.main()


if __name__ == '__main__':
    sys.exit(main())
```
To turn an entry point script into a PyInstaller package, just copy that code into its own script and run PyInstaller on it. Because `awscli` contains data files and some horrible dynamic importing you also need to force PyInstaller to collect its data files and all its submodules.
```fish
> cp /usr/local/bin/aws .
> pyinstaller --collect-submodules=awscli --collect-datas=awscli aws
> ./dist/aws/aws 
Note: AWS CLI version 2, the latest major version of the AWS CLI, is now stable and recommended for general use. For more information, see the AWS CLI version 2 installation instructions at: https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html

usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]
To see help text, you can run:

  aws help
  aws <command> help
  aws <command> <subcommand> help
```

**Reply by RobinFrcd** [reply]
Oh, that's nice to know. I didn't even bother to look at what was inside the bin/aws, thought it was a real binary indeed.

Thanks for your commands ! :pray:  I'll try to find a way to make it fully work with my current script which were calling the CLI like this:

```python
python_path = sys.executable.split(os.sep)
awscli_path = os.sep.join(python_path[:-1] + ["aws"])
process = subprocess.Popen(
    f"{awscli_path} s3 sync {src} {dst}",  
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE,
    shell=True,
    env=dict(os.environ, **env),
)
```

But I guess that's not possible from a PyInstall executable, right ?

**Reply by bwoodsend** [reply]
Correct, using subprocesses to call entry point scripts will not work. You're better off figuring out what code that entry point runs then using that in-process instead. That normally involves digging through the source code a bit. In this case we want to call `awscli.clidriver.main()` but that doesn't allow arguments so we can't use that so we find [its definition](https://github.com/aws/aws-cli/blob/1df6ee16f4472be368173f176e1c2c7a67e22dfb/awscli/clidriver.py#L68) and see if that's usable.
```python
def main():
    driver = create_clidriver()
    rc = driver.main()
    HISTORY_RECORDER.record('CLI_RC', rc, 'CLI')
    return rc
```
That `driver.main()` [optionally takes command line arguments](https://github.com/aws/aws-cli/blob/1df6ee16f4472be368173f176e1c2c7a67e22dfb/awscli/clidriver.py#L193) so that's the one to use.
```python
from awscli.clidriver import create_clidriver
awscli_driver = create_clidriver()
awscli_driver.main(["s3", "sync", src, dst])
```
---
P.S. Always use lists instead string commands to `Popen()`. i.e. `Popen([awscli_path, "s3", "sync", src, dst])` without `shell=True` instead of `f"{awscli_path} s3 sync {src} {dst}"` which will blow up if any of those variables contains spaces or special characters.

**Reply by RobinFrcd** [reply]
Again, thank you so much for your precious help ! :pray: 

Just to be sure, there's really no way to call the `python` executable from PyInstaller, it doesn't exist, right ? Seems like parsing stdout for [multiprocess is much harder than subprocess](https://stackoverflow.com/a/74072072/5692012), but if it's not possible to problem, I'll go this way.

Thanks again !

**Reply by bwoodsend** [reply]
> Just to be sure, there's really no way to call the python executable from PyInstaller, it doesn't exist, right.

It doesn't exist.

**Reply by bwoodsend** [reply]
I don't use any of these packages so I don't really know but you appear to trying to wrap a CLI tool into a Python API when that CLI tool is just a wrapper around another Python API ([boto3](https://pypi.org/project/boto3/)). Is cutting the middle CLI tool out and using boto3 directly not an option?

**Reply by RobinFrcd** [reply]
That was indeed my initial plan, but for some reason the boto team don't want to implement the sync function in the SDK: https://github.com/boto/boto3/issues/358

So I'm stuck with using shady workarounds :/
